sharktank/sharktank/tools/tuner/candidate_gen.py

```diff
@@ -0,0 +1,1386 @@
+
+
+tune_logger = logging.getLogger("tune")
+
+
+class DispatchKind(Enum):
+    conv = 1
+    mmt = 2
+    contraction = 3
+    batch_mmt = 4
+    batch_matmul = 5
+    broadcast_rhs_mmt = 6
+
+
+class ElementType(Enum):
+    i8 = 1
+    i32 = 2
+    f8 = 3
+    f16 = 4
+    f32 = 5
+
+    @property
+    def bitwidth(self) -> int:
+        match self:
+            case ElementType.i8 | ElementType.f8:
+                return 8
+            case ElementType.f16:
+                return 16
+            case ElementType.i32 | ElementType.f32:
+                return 32
+            case _:
+                assert False, "unhandled case"
+
+    def __str__(self) -> str:
+        return self.name
+
+
+@dataclass
+class ShapedType:
+    shape: list[int]
+    element_type: ElementType
+
+    def rank(self) -> int:
+        return len(self.shape)
+
+    @property
+    def bitwidth(self) -> int:
+        return self.element_type.bitwidth
+
+    def __str__(self) -> str:
+        dim_to_str = lambda dim: str(dim) if dim != -1 else "?"
+        return "x".join(map(dim_to_str, self.shape)) + "x" + str(self.element_type)
+
+
+@dataclass
+class MatmulSize:
+    M: int
+    N: int
+    K: int
+    B: int = 1
+
+
+@dataclass
+class ProblemSize:
+    matmul_size: MatmulSize
+    lhs_type: ShapedType
+    rhs_type: ShapedType
+    res_type: ShapedType
+    dispatch_kind: DispatchKind
+
+    @property
+    def MNK(self) -> tuple[int, int, int]:
+        return (self.matmul_size.M, self.matmul_size.N, self.matmul_size.K)
+
+
+@dataclass
+class MfmaIntrinsic:
+    input_type: ElementType
+    m: int
+    n: int
+    k: int
+    output_type: ElementType
+
+    def __str__(self) -> str:
+        input = str(self.input_type).upper()
+        output = str(self.output_type).upper()
+        return f"MFMA_{input}_{self.m}x{self.n}x{self.k}_{output}"
+
+    @staticmethod
+    def mfma_f16_16x16x16_f32():
+        return MfmaIntrinsic(ElementType.f16, 16, 16, 16, ElementType.f32)
+
+    @staticmethod
+    def mfma_f16_32x32x8_f32():
+        return MfmaIntrinsic(ElementType.f16, 32, 32, 8, ElementType.f32)
+
+    @staticmethod
+    def mfma_i8_16x16x32_i32():
+        return MfmaIntrinsic(ElementType.i8, 16, 16, 32, ElementType.i32)
+
+    @staticmethod
+    def mfma_i8_32x32x16_i32():
+        return MfmaIntrinsic(ElementType.i8, 32, 32, 16, ElementType.i32)
+
+    @staticmethod
+    def all():
+        return [
+            MfmaIntrinsic.mfma_f16_16x16x16_f32(),
+            MfmaIntrinsic.mfma_f16_32x32x8_f32(),
+            MfmaIntrinsic.mfma_i8_16x16x32_i32(),
+            MfmaIntrinsic.mfma_i8_32x32x16_i32(),
+        ]
+
+
+@dataclass
+class Configuration:
+    subgroup_size: int
+    workgroup_size: list[int]
+    intrinsic: MfmaIntrinsic
+    tile_sizes: list[int]
+    subgroup_m_count: int
+    subgroup_n_count: int
+    waves_per_eu: int
+
+
+class MlirRegex(str, Enum):
+    ssa_value = r"%[a-zA-Z0-9-_]+"
+    tensor_type = r"tensor<(([0-9]+x)+((f|i)[0-9]+))>"
```

Comment:
This looks tricky to reason about and might be missing coverage for some types like bf16 and more exotic types like `"f8e4m3fnuz"`. Is a regex matcher really needed? How is this used? If it's being used to parse through the input program, there should be some MLIR Python APIs you can use for programmatic manipulation instead of needing to parse through the format.

Reply:
The parsing / mutation is still text-based. Note that this parsing is pipeline-specific -- it only supports dispatches covered by `LLVMGPUVectorDistribution`. IMO this should be fine as long as we check that this is indeed the pipeline used, and pick a less general name like `VectorDistributionMlirRegex`. Longer term, we should definitely move all of the parsing to the mlir/iree bindings.

Would you be opposed to landing the first version that still depends on text processing for the input mlir?

Reply:
Added `LLVMGPUVectorDistribution` checks for now